src/PureScript/Backend/Optimizer/Tracer/Printer.purs

```diff
@@ -0,0 +1,516 @@
+        ]
+  D.lines
+    [ D.lines $ Array.mapWithIndex printCond (NonEmptyArray.toArray conds)
+    , D.text "else"
+    , D.indent $ printBackendExpr fallback
+    ]
+
+printBackendSyntax :: BackendSyntax BackendExpr -> Doc Void
+printBackendSyntax = case _ of
+  Var qi ->
+    printQualifiedIdent qi
+  Local mbIdent lvl ->
+    printLocal mbIdent lvl
+  Lit lit ->
+    printLiteral lit
+  App fn args ->
+    (NonEmptyArray.toArray args) # flip Array.foldl (printBackendExpr fn) \acc next ->
+      printCurriedApp acc $ printBackendExpr next
+  Abs args body -> do
+    let
+      printArg arg =
+        D.words
+          [ D.text "\\" <> uncurry printLocal arg
+          , D.text "->" <> D.space
+          ]
+    printIndentedParens' (foldMap1 printArg args) $ printBackendExpr body
+  UncurriedApp fn args ->
+    wrapInParens $ D.words $ map printBackendExpr $ Array.cons fn args
+  UncurriedAbs args body ->
+    printIndentedParens'
+      (D.text "\\" <> (D.words $ map (uncurry printLocal) args) <> D.space <> D.text "->")
+      $ printBackendExpr body
+  UncurriedEffectApp fn args ->
+    D.text "(#" <> (D.words $ map printBackendExpr $ Array.cons fn args) <> D.text "#)"
+  UncurriedEffectAbs args body ->
+    D.lines
+      [ D.text "(#\\" <> (D.words $ map (uncurry printLocal) args) <> D.space <> D.text "->"
+      , D.indent $ printBackendExpr body
+      , D.text "#)"
+      ]
+  Accessor expr accessor ->
+    printIndentedParens (printBackendExpr expr) <> printBackendAccessor accessor
+  Update expr propArr -> do
+    let sep = D.break <> D.text ", "
+    printIndentedParens $
+      D.lines
+        [ printBackendExpr expr
+        , D.indent $ D.lines
+            [ D.text "{"
+            , D.foldWithSeparator sep $ map (printProp printBackendExpr) propArr
+            , D.text "}"
+            ]
+        ]
+
+  CtorSaturated qi _ ctorName _ values ->
+    wrapInParens $
+      D.words
+        [ printQualifiedIdent qi <> D.text "." <> printProperName ctorName
+        , D.words $ map (printBackendExpr <<< snd) values
+        ]
+
+  CtorDef _ _ ctorName [] ->
+    printLet "let" (printIdent ctorName)
+      $ printRecord
+          [ (Prop "tag" $ ExprSyntax mempty $ Lit $ LitString $ (\(Ident i) -> i) ctorName) ]
+
+  CtorDef _ _ ctorName args ->
+    printLet "let" (printIdent ctorName)
+      $ printIndentedParens'
+          (guard (not $ Array.null args) $ D.words [ D.text "\\", D.words $ map D.text args, D.text "->" ])
+      $ printRecord
+      $ Array.cons (Prop "tag" $ ExprSyntax mempty $ Lit $ LitString $ (\(Ident i) -> i) ctorName)
+      $ args <#> \arg -> Prop arg $ ExprSyntax mempty $ Var $ Qualified Nothing $ Ident arg
+
+  LetRec lvl bindings body ->
+    D.lines
+      [ D.text "letrec"
+      , D.indent
+          $ D.lines
+          $ bindings <#>
+              ( \(Tuple ident binding) ->
+                  printLet "let" (printLocal (Just ident) lvl) $ printBackendExpr binding
+              )
```

Comment:
Perhaps this should be changed to output something like 
```
letrec-<lvl> identifier1 = ...
letrec-<lvl> identifier2 = ...
```
rather than its current output:
```
letrec
  let identifier1 = ...
  let identifier2 = ...
```